redis/protocol_parser.rl

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2019 pengjian.uestc @ gmail.com
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <memory>
+#include <iostream>
+#include <algorithm>
+#include <functional>
+#include "bytes.hh"
+#include "redis/request.hh"
+#include <seastar/core/ragel.hh>
+
+using namespace seastar;
+using namespace redis;
+%%{
+
+machine redis_resp_protocol;
+
+access _fsm_;
+
+action mark {
+    g.mark_start(p);
+}
+
+action start_blob {
+    g.mark_start(p);
+    _size_left = _arg_size;
+}
+
+action start_command {
+    g.mark_start(p);
+    _size_left = _arg_size;
+}
+
+action advance_blob {
+    auto len = std::min(static_cast<uint32_t>(pe - p), _size_left);
+    _size_left -= len;
+    p += len;
+    if (_size_left == 0) {
+      _req._args.push_back(str());
+      p--;
+      fret;
+    }
+    p--;
+}
+
+action advance_command {
+    auto len = std::min(static_cast<uint32_t>(pe - p), _size_left);
+    _size_left -= len;
+    p += len;
+    if (_size_left == 0) {
+      _req._command = str();
+      p--;
+      fret;
+    }
+    p--;
+}
+
+
+crlf = '\r\n';
+u32 = digit+ >{ _u32 = 0;}  ${ _u32 *= 10; _u32 += fc - '0';};
+args_count = '*' u32 crlf ${_req._args_count = _u32 - 1;};
+blob := any+ >start_blob $advance_blob;
+command := any+ >start_command $advance_command;
+arg = '$' u32 crlf ${ _arg_size = _u32;};
+
+main := (args_count (arg @{fcall command; } crlf) (arg @{fcall blob; } crlf)+) ${_req._state = request_state::ok;};
+
+prepush {
+    prepush();
+}
+
+postpop {
+    postpop();
+}
+
+}%%
+
+class redis_protocol_parser : public ragel_parser_base<redis_protocol_parser> {
+    %% write data nofinal noprefix;
```

Comment:
```
FAILED: build/release/redis/service.o 
g++ -MD -MT build/release/redis/service.o -MF build/release/redis/service.o.d -I/home/amos/scylladb.com/scylla-build-example/scylla-redis/seastar/include -I/home/amos/scylladb.com/scylla-build-example/scylla-redis/build/release/seastar/gen/include -std=gnu++17 -U_FORTIFY_SOURCE -Wno-maybe-uninitialized -march=westmere -Werror=unused-result -fconcepts -DSEASTAR_API_LEVEL=2 -DSEASTAR_HAVE_GCC6_CONCEPTS -DSEASTAR_USE_STD_OPTIONAL_VARIANT_STRINGVIEW -DSEASTAR_HAVE_DPDK -DFMT_SHARED -I/usr/include/p11-kit-1 -ffile-prefix-map=/home/amos/scylladb.com/scylla-build-example/scylla-redis=.  -march=westmere -DBOOST_TEST_DYN_LINK -DTHRIFT_USES_BOOST  -fvisibility=hidden -Wall -Werror -Wno-maybe-uninitialized -Wno-tautological-compare -Wno-missing-braces -Wno-misleading-indentation -Wno-overflow -Wno-noexcept-type -Wno-nonnull-compare -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-stringop-overflow -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DHAVE_LIBSYSTEMD=1 -DHAVE_LZ4_COMPRESS_DEFAULT -O3 --param inline-unit-growth=300 -gz  -g -I. -I build/release/gen  -c -o build/release/redis/service.o redis/service.cc
In file included from ./redis/server.hh:39,
                 from redis/service.cc:24:
build/release/gen/redis/protocol_parser.hh:43:20: error: in-class initialization of static data member 'const char redis_protocol_parser::_nfa_targs []' of incomplete type
   43 |  static const char _nfa_targs[] = {
      |                    ^~~~~~~~~~
build/release/gen/redis/protocol_parser.hh:47:20: error: in-class initialization of static data member 'const char redis_protocol_parser::_nfa_offsets []' of incomplete type
   47 |  static const char _nfa_offsets[] = {
      |                    ^~~~~~~~~~~~
build/release/gen/redis/protocol_parser.hh:53:20: error: in-class initialization of static data member 'const char redis_protocol_parser::_nfa_push_actions []' of incomplete type
   53 |  static const char _nfa_push_actions[] = {
      |                    ^~~~~~~~~~~~~~~~~
build/release/gen/redis/protocol_parser.hh:57:20: error: in-class initialization of static data member 'const char redis_protocol_parser::_nfa_pop_trans []' of incomplete type
   57 |  static const char _nfa_pop_trans[] = {
      |                    ^~~~~~~~~~~~~~
```

Reply:
Hi @amoskong , what's the version of ragel which you used?

I use the ragel with version 6.8, and there are no errors as you mentioned.

> Ragel State Machine Compiler version 6.8 Feb 2013
> Copyright (c) 2001-2009 by Adrian Thurston

Reply:
I built by `./tools/toolchain/dbuild ./reloc/build_reloc.sh`.

[scylla-redis]$ ./tools/toolchain/dbuild -it -- rpm -q ragel
ragel-7.0.0.11-3.fc30.x86_64

Reply:
I found that the same problem was reported in this issue: https://github.com/scylladb/seastar/issues/296

Reply:
I change the ragel to latest version (Ragel State Machine Compiler version 7.0.0.12 May 2019
). The errors are still exists.
But the ragel with version 6.8 works fine.

Can we change the ragel version to 6.8 ?

Reply:
> I change the ragel to latest version (Ragel State Machine Compiler version 7.0.0.12 May 2019). The errors are still exists. But the ragel with version 6.8 works fine.
> 
> Can we change the ragel version to 6.8 ?

/CC @avikivity ?

Reply:
The building completed without error after downgraded ragel to 6.8-5

#### The downgrade command:
```
sudo dnf downgrade -y https://rpmfind.net/linux/mageia/distrib/5/x86_64/media/core/release/ragel-6.8-5.mga5.x86_64.rpm
```

#### New docker image with ragel-6.8-5
```
diff --git a/tools/toolchain/image b/tools/toolchain/image
index 3d89f8f..2e9992e 100644
--- a/tools/toolchain/image
+++ b/tools/toolchain/image
@@ -1 +1 @@
-docker.io/scylladb/scylla-toolchain:fedora-30-20190911
+docker.io/amoskong/scylla-toolchain:fedora-30-ragel6.8
```

Reply:
I don't follow the details of this discussion (I'm not at all familiar with Ragel or what changed in its versions), but please don't make your code rely on an antique version of Ragel. My build machine has ragel 7.0.0.11 and I don't have a problem building Seastar of Scylla. The new code needs to work also on the modern Ragel 7.0 - it can't just work on Ragel 6.8. That's not a viable workaround, in my opinion.